internal/handlers/handlers.go

```diff
@@ -61,23 +84,168 @@ func shortURL(url string) string {
+
+			var errConflict *storage.URLConflictError
+			if errors.As(err, &errConflict) {
+				shortID = errConflict.URL
+
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(http.StatusConflict)
+			} else {
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+		} else {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusCreated)
+		}
+
+		resp := fmt.Sprintf("%s/%s", config.Options.FlagShortURL, shortID)
+
+		shortenResponse := models.ShortenResponse{
+			Result: resp,
+		}
+
+		enc := json.NewEncoder(w)
+		if err := enc.Encode(shortenResponse); err != nil {
+			logger.Log.Debug("Ошибка создания ответа", zap.Error(err))
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		logger.Log.Debug("sending HTTP 200 response")
+	}
+
+	return http.HandlerFunc(fn)
+}
 
-	if id == "" {
-		w.WriteHeader(http.StatusBadRequest)
-		return
+func (h *Handlers) SaveBatch() http.HandlerFunc {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		var batchReq []models.BatchShortenRequest
+
+		err := json.NewDecoder(r.Body).Decode(&batchReq)
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		var shortURLs []models.StorageURL
+
+		for _, value := range batchReq {
+			hashID, err := h.storage.GetURL(value.OriginalURL)
+			if err != nil {
+				logger.Log.Debug("error get url response", zap.Error(err))
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+
+			if hashID == "" {
+				hashID = shortURL(value.OriginalURL)
+				shortURLs = append(shortURLs, models.StorageURL{
+					OriginalURL: value.OriginalURL,
+					ShortURL:    hashID,
+				})
+			} else {
+				shortURLs = append(shortURLs, models.StorageURL{
+					OriginalURL: value.OriginalURL,
+					ShortURL:    hashID,
+				})
+			}
+		}
+
+		urls, err := h.storage.SaveBatchURL(r.Context(), shortURLs)
+		if err != nil {
+			logger.Log.Debug("error urls save", zap.Error(err))
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		res := make([]models.BatchShortenResponse, 0, len(urls))
+
+		for i, shortURL := range urls {
+			res = append(res, models.BatchShortenResponse{
+				CorrelationID: batchReq[i].CorrelationID,
+				ShortURL:      fmt.Sprintf("%s/%s", config.Options.FlagShortURL, shortURL),
+			})
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		//w.Header().Set("Location", config.Options.FlagShortURL)
+		w.WriteHeader(http.StatusCreated)
+
+		enc := json.NewEncoder(w)
+		if err := enc.Encode(res); err != nil {
+			logger.Log.Debug("error encoding response", zap.Error(err))
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		logger.Log.Debug("sending HTTP 200 response")
 	}
 
-	if fullURL, ok := urlStore[id]; ok {
-		http.Redirect(w, r, fullURL, http.StatusTemporaryRedirect)
-		return
+	return http.HandlerFunc(fn)
+}
+
+func (h *Handlers) PingDB() http.HandlerFunc {
+	fn := func(w http.ResponseWriter, r *http.Request) {
```

Comment:
 Так как `HandlerFunc` и есть `func(ResponseWriter, *Request)` не обязательно оборачивать функцию, а можно ее сразу вернуть.
```suggestion
	return  func(w http.ResponseWriter, r *http.Request) {
```